components/ScriptLibrary/include/LHVMTypes.h

```diff
@@ -0,0 +1,371 @@
+/******************************************************************************
+ * Copyright (c) 2018-2024 openblack developers
+ *
+ * For a complete list of all authors, please refer to contributors.md
+ * Interested in contributing? Visit https://github.com/openblack/openblack
+ *
+ * openblack is licensed under the GNU General Public License version 3.
+ *******************************************************************************/
+
+#pragma once
+
+#include <cassert>
+#include <cstdint>
+
+#include <array>
+#include <functional>
+#include <map>
+#include <string>
+
+namespace openblack::lhvm
+{
+
+static const std::array<std::string, 31> k_OpcodeNames = {
+    "END",    "JZ",   "PUSH", "POP",       "ADD",       "SYS",        "SUB",       "NEG",  "MUL", "DIV", "MOD",
+    "NOT",    "AND",  "OR",   "EQ",        "NE",        "GE",         "LE",        "GT",   "LT",  "JMP", "SLEEP",
+    "EXCEPT", "CAST", "RUN",  "ENDEXCEPT", "RETEXCEPT", "FAILEXCEPT", "BRKEXCEPT", "SWAP", "LINE"};
+
+enum class LHVMVersion : uint32_t
+{
+	BlackAndWhite = 7,
+	CreatureIsle = 8,
+	BlackAndWhite2 = 12,
+
+	Last = 0xFFFFFFFF
+};
+
+enum class DataType : uint32_t
+{
+	None = 0,
+	Int,
+	Float,
+	Vector,
+	Object,
+	Unk5,
+	Boolean,
+	Unk7,
+
+	Last = 0xFFFFFFFF
+};
+static_assert(sizeof(DataType) == 4);
+
+static const std::array<std::string, 8> k_DataTypeNames = {"NONE",   "INT",  "FLOAT",   "VECTOR",
+                                                           "OBJECT", "UNK2", "BOOLEAN", "UNK4"};
+
+static const std::array<std::string, 8> k_DataTypeChars = {"", "I", "F", "V", "O", "", "B", ""};
+
+class VMValue
+{
+public:
+	VMValue()
+	    : floatVal(0.0f)
+	{
+	}
+
+	explicit VMValue(float value)
+	    : floatVal(value)
+	{
+	}
+
+	explicit VMValue(int32_t value)
+	    : intVal(value)
+	{
+	}
+
+	explicit VMValue(uint32_t value)
+	    : uintVal(value)
+	{
+	}
+
+	union
+	{
+		float floatVal;
+		int32_t intVal;
+		uint32_t uintVal;
+	};
+};
+static_assert(sizeof(VMValue) == 4);
+
+class VMVar
+{
+public:
+	VMVar(DataType type, VMValue value, std::string name)
+	    : type(type)
+	    , value(value)
+	    , name(std::move(name))
+	{
+	}
+
+	DataType type;
+	union
+	{
+		VMValue value;
+		float floatVal;
+		int32_t intVal;
+		uint32_t uintVal;
+	};
+	std::string name;
+};
+
+class VMStack
+{
+public:
+	VMStack() = default;
+
+	uint32_t count {0};
+	std::array<VMValue, 32> values {};
+	std::array<DataType, 32> types {};
+	uint32_t pushCount {0};
+	uint32_t popCount {0};
+};
+static_assert(sizeof(VMStack) == 268);
+
+enum class ScriptType : uint32_t
```

Comment:
This is a kind of bit field. I say "kind of" because actually a script can have just one type assigned (i.e. only 1 bit set). The reason why it uses non-overlapping-bit values is that there are some functions which must be able to operate on multiple types at once.

For example in `LHVM::StopTasksOfType(ScriptType typesMask)` you can say "stop all tasks of types TempleHelp and TempleSpecial, but not the others" (I guess that this could be helpful when exiting the temple).
Another example is `LHVM::LookIn(ScriptType allowedScriptTypesMask)` which is used to run a virtual CPU turn; here you want to run just MultiplayerHelp scripts if you are in skirmish, or all but Temple* and MultiplayerHelp if you are in story mode.

Where should I put the `Flag` word, in the type name, in each entry name, or both?